### DISCUSSION TITLE
Any plan to introduce arm64e (with an 'e') target on macOS?

### DISCUSSION THREAD

**Discussion by lelegard** [Homebrew/discussions#4217] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
-
```

### Output of `brew doctor`

```shell
-
```

### Description of issue

Hi Homebrew maintainers,

Is there any strategy for the introduction of the arm64e (with an 'e') platform on Homebrew for macOS on Apple Silicon?

Currently, on macOS for Apple Silicon, arm64 (without 'e') is the default target for the clang compiler. The arm64e ABI is currently in "preview mode" and not supported by default for third-party applications. 

As of macOS 13 Ventura, third-party applications (outside the protected system enclave) which are built for arm64e are not allowed to run. Technically, they can run but they are not allowed to, by default. You need to set a specific boot flag and disable the system integrity to test them. I have explained [my experimentations with arm64e on macOS in a Gist here](https://gist.github.com/lelegard/009cbdae78e5993ed9e02160b9130d7f), if anyone is interested. This is just for tests.

However, this will change and arm64e will become the default platform, or at least a supported one. I do not know when. However, all Apple-provided applications in the protected system enclave (in `/bin` and `/usr/bin`) as well as all libraries in the system library cache are already built for arm64e.

As explained in the above-mentioned Gist, arm64 applications can call arm64e libraries but not the opposite. This is normal for security reasons. Indeed, the main benefit of arm64e target is the advanced security in the code, typically to mitigate malware code injections through ROP and JOP attacks.

Currently, all Homebrew binaries are built for arm64. At some point, Homebrew will need to support arm64e while continuing to support arm64 on previous macOS versions. I wonder if there is any plan for this.

Technically speaking, Homebrew could remain in arm64 mode since those applications will continue to be supported on arm64e platforms. However, there are at least two reasons not to do this, a weak one and a strong one:

1. The weak one: That would be a pity not to take advantage of the increased code security in arm64e.
2. The strong one: Homebrew contains a lot of libraries which are used by users' applications. For the reason I mentioned above, it is not possible for a user to build his applications for arm64e if the libraries it uses from Homebrew are built for arm64.

For instance, in my experimentations, after rebooting my M1 Mac with arm64e enabled, I could not rebuild my application (@tsduck) for this reason. When disabling all dependencies to Homebrew-provided libraries, the application builds correctly in arm64e mode and all tests succeed. Do not misunderstand me, I am not complaining, this is normal as of today. I am just thinking about tomorrow.

Could you please share your thoughts about this?
Thanks
-Thierry

**Comment by SMillerDev** [comment]
Once it becomes the default I expect we'll use this default. Until then, it seems like a lot of work.

**Comment by Bo98** [comment, marked as answer]
> As of macOS 13 Ventura, third-party applications (outside the protected system enclave) which are built for arm64e are not allowed to run. Technically, they can run but they are not allowed to, by default. You need to set a specific boot flag and disable the system integrity to test them. I have explained [my experimentations with arm64e on macOS in a Gist here](https://gist.github.com/lelegard/009cbdae78e5993ed9e02160b9130d7f), if anyone is interested. This is just for tests.

You've summarised well here why we don't have support for arm64e. Apple explicitly recommend not building for arm64e yet (except kernel extensions), as the ABI is not yet stable or finalised.

> Currently, all Homebrew binaries are built for arm64. At some point, Homebrew will need to support arm64e while continuing to support arm64 on previous macOS versions. I wonder if there is any plan for this.

Any arm64e introduction will not happen overnight. I would expect time, likely a minimum of a few months between a WWDC announcement and a major macOS release around October some year.

Really most of the work is individual package compatibility, which at this point is more a question for the developers of each software.

The main bit for us is choosing what exactly we ship which we'll do in that time of a beta being available. Exactly what we do depends on how exactly it ends up working (e.g. could we get away with providing only arm64e for that version of macOS?).

Even the compatibility side gets fuzzed when you have system Ruby (arm64e) load user-installed arm64 bundles.

There's not much point spending effort on theoreticals before there's something concrete in place.

**Reply by vladak** [reply]
One potentially interesting tidbit: when attempting to use `stdbuf` from Homebrew with system utilities such as `/bin/bash`, `/usr/bin/env` or such which are built as x86_64+amd64e (no plain amd64) fat binaries, this fails with:
```
prefix = stdbuf -o0

dyld[3909]: terminating because inserted dylib '/opt/homebrew/Cellar/coreutils/9.7/libexec/coreutils/libstdbuf.so' could not be loaded: tried: '/opt/homebrew/Cellar/coreutils/9.7/libexec/coreutils/libstdbuf.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e')), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/coreutils/9.7/libexec/coreutils/libstdbuf.so' (no such file), '/opt/homebrew/Cellar/coreutils/9.7/libexec/coreutils/libstdbuf.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e'))
dyld[3909]: tried: '/opt/homebrew/Cellar/coreutils/9.7/libexec/coreutils/libstdbuf.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e')), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/coreutils/9.7/libexec/coreutils/libstdbuf.so' (no such file), '/opt/homebrew/Cellar/coreutils/9.7/libexec/coreutils/libstdbuf.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e'))
./test.sh: line 8:  3909 Abort trap: 6           $prefix /bin/ls
```
This is because `stdbuf` works by preloading the `libstdbuf.so` library which calls `setvbuf()` in its init code with desired settings before executing the target program.

I bumped into this with the Github macOS runners with (virtualized, presumably M1 based) macOS 14.7.5 (and also 15), however cannot reproduce this locally on my company issued Macbook M3 pro running macOS 15.4.1 for some reason although the binaries have the same high level properties. Possibly there is something about the environment or setup of these machines.

Anyhow, my use case was to apply `stdbuf` on bunch of shell scripts starting with `#!/usr/bin/env bash`. Given the restrictions, the only cure is to rebuild coreutils with amd64e, I think.

**Reply by lelegard** [reply]
@vladak 

The case of stdbuf is very different. This is a matter of SIP (System Integrity Protection), not a matter of target architecture, arm64 vs. arm64e.

Stdbuf is a functional / performance tool but a security nightmare. It is designed to inject a 3rd-party shared library into applications. If the application is a system tool inside the SIP perimeter, then the OS _must_ refuse the injection of an external shared library (outside SIP). Otherwise, it would invalidate the whole principle of SIP.
